File: coding/src/com/algos/li02_hashing/L290_WordPattern.java
package  com.algos.li02_hashing;

import java.util.HashMap;

public class L290_WordPattern {

	public static void main(String[] args) {
		L290_WordPattern obj = new L290_WordPattern();
		
		System.out.println(obj.wordPattern("abba", "dog cat cat dog"));
	}
	
	/*
	 * Problem: https://leetcode.com/problems/word-pattern/
	 * Companies: Amazon
     * Idea: Maintain a mapping between char to word and vice versa
	 * */
	public boolean wordPattern(String pattern, String s) {
        HashMap<Character, String> charToWord = new HashMap<>();
        HashMap<String, Character> wordToChar = new HashMap<>();

        char[] chars = pattern.toCharArray();
        String[] words = s.split(" ");
        
        if(chars.length != words.length) return false;
        
        for(int i=0; i<chars.length; i++){
            if(charToWord.containsKey(chars[i]) && !charToWord.get(chars[i]).equals(words[i]))
                return false;                    
            if(wordToChar.containsKey(words[i]) && wordToChar.get(words[i]) != chars[i])
                return false;                    
            
            charToWord.put(chars[i], words[i]);
            wordToChar.put(words[i], chars[i]);
        }
        
        return true;
    }

}

File: dsa/concepts/02_hashing/python/l290_word_pattern.py
"""
L290 - Word Pattern

Problem: https://leetcode.com/problems/word-pattern/
Idea: Maintain a mapping between char to word and vice versa
Time: O(n)
Space: O(n)
"""

def word_pattern(pattern, s):
    char_to_word = {}
    word_to_char = {}
    
    words = s.split()
    
    if len(pattern) != len(words):
        return False
    
    for char, word in zip(pattern, words):
        # Check if char already mapped to a different word
        if char in char_to_word and char_to_word[char] != word:
            return False
        
        # Check if word already mapped to a different char
        if word in word_to_char and word_to_char[word] != char:
            return False
        
        # Create the mappings
        char_to_word[char] = word
        word_to_char[word] = char
    
    return True


if __name__ == "__main__":
    test_cases = [
        ("abba", "dog cat cat dog"),
        ("abba", "dog cat cat fish"),
        ("aaaa", "dog cat cat dog"),
        ("abba", "dog dog dog dog")
    ]
    
    for i, (pattern, s) in enumerate(test_cases, 1):
        print(f"Test case {i}: pattern = '{pattern}', s = '{s}'")
        print(f"Word pattern match: {word_pattern(pattern, s)}")
        print()
